# Classifying Boolean Values
Michaela Webster - mawebster9

Before we begin anything, we need to import all of the libraries/functions that we will be using throughout our notebook. Pandas is a crucial part of this process since it houses our data. Another important feature of this process is the scikit-learn library - this library is our one-stop-shop for our machine learning needs. 

In [90]:
#our data structure
import pandas as pd

#bag of words vectorizer - take inverse frequency of words to assign weights
from sklearn.feature_extraction.text import TfidfVectorizer

#split data into training/test data, validate our models, and specify number of folds for training/test data
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold 

# our 5 classification models
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier

## 1. Import Data and Set X & y

For this example, we will focus on all of the boolean fields. The boolean fields are all of the true/false values and can account for a lot of different attributes that contain a lot of useful data surrounding a case. Our goal is to see if the true/false fields can be used to determine whether a case should be denied or granted with high accuracy. For this, our X is all of the true/false fields and its associated label, or y, is the denied field.

In [91]:
#connect to CSV file that contains our data
path_to_file = "https://raw.githubusercontent.com/mawebster9/ThesisCode/master/appeals_query.csv"

#open, read, and store our data into a pandas dataframe
appeals_data = pd.read_csv(path_to_file, encoding='latin-1')

In [92]:
#assign our attributes to X (all boolean fields) and y
X = appeals_data.select_dtypes('bool')
y = X['Denied']

In [93]:
#remove the 'Denied' field from the X data frame
X = X.drop('Denied',axis=1)

In [94]:
#print the first record in X to verify the previous step
X.iloc[0]

Smith                     False
Female                    False
Position_Eligibility      False
No_Falsification          False
Rebut_Falsification       False
Falsification(s)           True
Domestic_Violence         False
Previous_Clearance        False
Traumatic_Life_Event      False
Caused_Death              False
Child_Sexual_Abuse        False
Child_Pornography         False
Prostitutes               False
Fmr_Military_LawE         False
Adverse_Affirmed          False
Favorable_Affirmed        False
Granted                   False
Failed_to_Mitigate        False
Success_to_Mitigate       False
Adverse_Reversed          False
Revoked_Fav_Reversed      False
Adverse_Remanded          False
Favorable_Remanded        False
Remanded_wInstructions    False
Recommend_Waiver          False
Decision_Other            False
Decision_Unknown          False
Security_Violations       False
Foreign_Influence         False
Foreign_Preference        False
Sexual_Behavior           False
Personal

## 2. Set-up X: Fix Boolean Values

We need to set up our X values in a way that a machine learning algorithm can understand it. In order to do this we need to change our true/false values into a numeric format. To do this, we need to change all true values to 1.0 and all false values to 0.0.

In [95]:
#replace all instances: True = 1.0, False = 0.0
X.columns.tolist()
for i in (X.columns.tolist()):
    X[i] = X[i].replace(True,1)

In [96]:
#print out top 10 records in X to ensure True was changed to 1.0 and False was changed to 0.0
X.head(10)

Smith  Female  Position_Eligibility  No_Falsification  Rebut_Falsification  \
0    0.0     0.0                   0.0               0.0                  0.0   
1    0.0     1.0                   0.0               1.0                  0.0   
2    0.0     0.0                   0.0               1.0                  0.0   
3    0.0     0.0                   0.0               1.0                  0.0   
4    0.0     0.0                   0.0               0.0                  0.0   
5    0.0     0.0                   0.0               1.0                  0.0   
6    0.0     0.0                   0.0               0.0                  0.0   
7    0.0     0.0                   0.0               1.0                  0.0   
8    0.0     0.0                   0.0               1.0                  0.0   
9    0.0     0.0                   0.0               0.0                  0.0   

   Falsification(s)  Domestic_Violence  Previous_Clearance  \
0               1.0                0.0                 0.0   
1               0.0                0.0                 0.0   
2               0.0                0.0                 0.0   
3               0.0                0.0                 0.0   
4               1.0                0.0                 0.0   
5               0.0                0.0                 0.0   
6               1.0                0.0                 0.0   
7               0.0                0.0                 0.0   
8               0.0                0.0                 0.0   
9               1.0                0.0                 1.0   

   Traumatic_Life_Event  Caused_Death  ...  Alcohol  Drugs  Emotional_Mental  \
0                   0.0           0.0  ...      0.0    1.0               0.0   
1                   0.0           0.0  ...      0.0    1.0               0.0   
2                   0.0           0.0  ...      0.0    1.0               0.0   
3                   0.0           0.0  ...      0.0    1.0               0.0   
4                   0.0           0.0  ...      1.0    1.0               0.0   
5                   0.0           0.0  ...      1.0    0.0               0.0   
6                   0.0           0.0  ...      0.0    1.0               0.0   
7                   0.0           0.0  ...      1.0    0.0               0.0   
8                   0.0           0.0  ...      0.0    0.0               0.0   
9                   0.0           0.0  ...      0.0    1.0               0.0   

   Criminal_Conduct  Handling_PI  Outside_Activities  Use_InfoSys  Deception  \
0               1.0          0.0                 0.0          0.0        0.0   
1               1.0          0.0                 0.0          0.0        0.0   
2               0.0          0.0                 0.0          0.0        0.0   
3               0.0          0.0                 0.0          0.0        0.0   
4               1.0          0.0                 0.0          0.0        0.0   
5               0.0          0.0                 0.0          0.0        0.0   
6               1.0          0.0                 0.0          0.0        0.0   
7               0.0          0.0                 0.0          0.0        0.0   
8               1.0          0.0                 0.0          0.0        0.0   
9               1.0          0.0                 0.0          0.0        0.0   

   CAC  Unknown_Guideline  
0  0.0                0.0  
1  0.0                0.0  
2  0.0                0.0  
3  0.0                0.0  
4  0.0                0.0  
5  0.0                0.0  
6  0.0                0.0  
7  0.0                0.0  
8  0.0                0.0  
9  0.0                0.0  

[10 rows x 43 columns]

In [97]:
#Alternatively, you could print the first record in X
X.iloc[0]

Smith                     0.0
Female                    0.0
Position_Eligibility      0.0
No_Falsification          0.0
Rebut_Falsification       0.0
Falsification(s)          1.0
Domestic_Violence         0.0
Previous_Clearance        0.0
Traumatic_Life_Event      0.0
Caused_Death              0.0
Child_Sexual_Abuse        0.0
Child_Pornography         0.0
Prostitutes               0.0
Fmr_Military_LawE         0.0
Adverse_Affirmed          0.0
Favorable_Affirmed        0.0
Granted                   0.0
Failed_to_Mitigate        0.0
Success_to_Mitigate       0.0
Adverse_Reversed          0.0
Revoked_Fav_Reversed      0.0
Adverse_Remanded          0.0
Favorable_Remanded        0.0
Remanded_wInstructions    0.0
Recommend_Waiver          0.0
Decision_Other            0.0
Decision_Unknown          0.0
Security_Violations       0.0
Foreign_Influence         0.0
Foreign_Preference        0.0
Sexual_Behavior           0.0
Personal_Conduct          1.0
Financial                 0.0
Alcohol   

## 3. Set-up y: Fix Boolean Values

We have our X values set up in a way that a machine learning algorithm can understand it, but now we need to fix our y values. In order to do this we need to change our true/false values into a numeric format. To do this, we need to change all true values to 1.0 and all false values to 0.0.

In [98]:
#replace: True = 1.0, False = 0.0
y = y.replace(True,1)

In [99]:
#print out counts for all y records - ensure that our replace statement worked
y.value_counts()

1.0    10862
0.0     9652
Name: Denied, dtype: int64

In [100]:
#Alternatively, you could print the first record in y
y.iloc[0]

1.0

### Final Data Check

Now that our X and y are in the right format, we need to ensure one last time that the dimensions of each dataframe are correct. For our X, we see that there are 20,514 rows and 43 columns (different attributes in dataset). For our y, we see that there are 20,514 rows and no columns.

Our data has passed the check and is ready to be used.

In [101]:
X.shape

(20514, 43)

In [102]:
y.shape

(20514,)

## 4. Machine Learning Step-by-Step

#### A. Run train_test_split() on X & y

This step is not needed for this notebook but it shows you how the train_test_split function works. Our X and y are randomly split up into training and testing groups. In this case, our test group will be comprised of 33% of the X data(test_size), and will be the same each time we run this line (random_state).

In [103]:
#break 33% of X and y into X_test and y_test, break other remaining 67% into X_train and y_train
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [104]:
#print size of training data (13,744/20,514 = .67)
X_train.shape

(13744, 43)

In [105]:
#print size of test data (6,770/20,514 = .33)
X_test.shape

(6770, 43)

#### B. Run fit() on X_train & y_train

The next step in our machine learning model is taking our training data and feeding it into an algorithm to build a model. This is essentially the step that teaches an algorithm that for each record X = y. To do this, there are a number of classification models. For this example we will focus on the LogisticRegression classifier.

In [106]:
#specify which classifier to use and set parameters
clf = LogisticRegression(random_state=0, solver='liblinear')

In [107]:
#send X and y into our classifier to build a model
logreg_model = clf.fit(X, y)

In [108]:
print(logreg_model)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=0, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)


#### C. Run predict() on X_test

The next step in our model is to take the model we just made using the training data and feeding the test data into it. This will output an array of values that the algorithm has determined to be the denied status.

In [109]:
#send our test data into the model we just created
y_pred = logreg_model.predict(X_test)

In [110]:
#print our results for the predictions
print("Here is the model's predictions: ")
print(y_pred)

Here is the model's predictions: 
[0. 1. 0. ... 0. 0. 1.]


#### D. Verify Accuracy of Model

Now that we have split our data into training and testing groups, created a model using a machine learning algorithm, and used the model to predict outcomes for our test data, it is time to verify how well our model did compared to the actual outcomes. To do this, there are a number of accuracy metrics. For this example we will focus on the accuracy score.

In [111]:
#import our score functions
from sklearn.metrics import accuracy_score

In [112]:
#compare y_test values with the predicted y values
score = accuracy_score(y_test, y_pred).mean()
print("Accuracy score for LogisticRegression classifier:  ", score)

Accuracy score for LogisticRegression classifier:   0.9957163958641063


***Here we can see that our LogisticRegression model was correct 99.57% of the time when predicting the denied status.***

## 5. Test for Best Classifier to Use

Now that we understand how machine learning is done, we can determine which model is the best choice for our data. In this example we will use  different classifiers and evaluate each against 5 accuracy metrics.

Note: the cross_val_score() function handles test_train_split(X,y,test_size=33,random_state-42), fit(X_train,y_train), predict(X_test), and also any of the accuracy score metrics. This function is essentially and all-in-one function.

In [113]:
classifiers = [RandomForestClassifier(n_estimators=5), GaussianNB(), LogisticRegression(solver='liblinear'), DecisionTreeClassifier(criterion='gini'), KNeighborsClassifier(n_neighbors=6)]
clf_names = ['RandomForest','GausianNB','LogisticRegression','DecisionTreeClassRegressor', 'KNeighbors']
metric_names = ['roc_auc','f1','accuracy','precision','recall']

scv = StratifiedKFold(n_splits=3)

scores_df = pd.DataFrame(index=metric_names,columns=clf_names)
clf_scores = []
for clf, name in zip(classifiers, clf_names):
    print('-----------------------------------------------------------------------------------------------------------')
    print('Classifier: ',clf)
    print('')
    print("Scoring Metrics: ")
    for metric in metric_names:
        score = cross_val_score(clf,X,y,scoring=metric, cv=scv).mean()
        clf_scores.append(score)
        print('\t*',metric,'score: ', score)
    scores_df[name] = clf_scores
    clf_scores = []

-----------------------------------------------------------------------------------------------------------
Classifier:  RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=5,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

Scoring Metrics: 
	* roc_auc score:  0.9970754501239952
	* f1 score:  0.9908797234769704
	* accuracy score:  0.9905914272360826
	* precision score:  0.992444556801408
	* recall score:  0.9912538786686828
-----------------------------------------------------------------------------------------------------------
Classifier:  GaussianNB(priors=None, var_smoothing=1e-09)

Scoring Metrics: 


## Final Results

For boolean classification, the top performing machine learning algorithm is RandomForrestClassifier followed closely behind by the DecisionTreeClassifier. Both of these classifiers work by building out a tree via simple if then else rules - the deeper the tree, the more complex the model.